# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/loader/ModuleLoadRequest.cpp
# Commit: c48ad530cfe2
# Full Hash: c48ad530cfe2161cd83e1fe52821d921a55fbae1
# Author: Yoshi Cheng-Hao Huang <allstars.chh@gmail.com>
# Date: 2025-09-24 18:03 +0000
# Regressor Bug: 1968890
# File Overlap Count: 3
# Overlapping Files: js/loader/ModuleLoadRequest.h, js/loader/ModuleLoadRequest.cpp, js/loader/ModuleLoaderBase.cpp
# Description:
#   Bug 1968890 - Part 5: Add ModuleLoadRequest::SetErroredLoadingImports(). r=jonco
# ==============================================================================

diff -r 42da5cac3411 -r c48ad530cfe2 js/loader/ModuleLoadRequest.cpp
--- a/js/loader/ModuleLoadRequest.cpp	Wed Sep 24 18:03:29 2025 +0000
+++ b/js/loader/ModuleLoadRequest.cpp	Wed Sep 24 18:03:30 2025 +0000
@@ -55,6 +55,7 @@
                         aFetchOptions, aIntegrity, aReferrer, aContext),
       mKind(aKind),
       mModuleType(aModuleType),
+      mErroredLoadingImports(false),
       mLoader(aLoader),
       mRootModule(aRootModule) {
   MOZ_ASSERT(mLoader);
@@ -65,7 +66,12 @@
 }
 
 bool ModuleLoadRequest::IsErrored() const {
-  return !mModuleScript || mModuleScript->HasParseError();
+  if (!mModuleScript || mErroredLoadingImports) {
+    return true;
+  }
+
+  MOZ_ASSERT_IF(mModuleScript->HasErrorToRethrow(), !IsDynamicImport());
+  return mModuleScript->HasParseError() || mModuleScript->HasErrorToRethrow();
 }
 
 void ModuleLoadRequest::Cancel() {
@@ -136,13 +142,7 @@
   }
 
   MOZ_ASSERT(!IsFinished());
-
-  mozilla::DebugOnly<bool> hasRethrow =
-      mModuleScript && mModuleScript->HasErrorToRethrow();
-
-  // When LoadRequestedModules fails, we will set error to rethrow to the module
-  // script and call ModuleErrored().
-  MOZ_ASSERT(IsErrored() || hasRethrow);
+  MOZ_ASSERT(IsErrored());
 
   if (IsFinished()) {
     // Cancelling an outstanding import will error this request.